CanonicalABI.md

```diff
@@ -0,0 +1,1082 @@
+elements in-memory. For this we can first define the `size` and `align` of all
+types:
+
+```
+def size(direction, type):
+  match type
+    f32 => 4
+    f64 => 8
+    s8 => 1
+    u8 => 1
+    s16 => 2
+    u16 => 2
+    s32 => 4
+    u32 => 4
+    s64 => 8
+    u64 => 8
+    char => 4
+    string |
+    list $ty => size(record([i32, i32]))
+    bitflags $fields => align_to($fields.len(), 32) * 4
+    record $fields =>
+      s = 0
+      for field in $fields
+        s = align_to(s, align(direction, field))
+        s += size(direction, field)
+      align_to(s, align(direction, record $fields))
+    variant $cases =>
+      s = 0
+      for case in $cases
+        if case:
+          s = max(s, size(direction, record([i32, case])))
+        else:
+          s = max(s, size(direction, i32))
+      s
+    handle => size(i32)
+    push-buffer | pull-buffer =>
+      if direction == "import":
+        size(record([i32, i32, i32]))
+      else:
+        size(i32)
+
+def align(direction, type):
+  match type
+    f32 = 4
+    f64 = 8
+    s8 = 1
+    u8 = 1
+    s16 = 2
+    u16 = 2
+    s32 = 4
+    u32 = 4
+    s64 = 8
+    u64 = 8
+    char = 4
+    string = 4
+    list $ty = 4
+    bitflags $fields = 4
+    record $fields =
+      a = 1
+      for field in $fields
+        a = max(a, align(direction, field))
+      a
+    variant $cases =
+      a = 4
+      for case in $cases
+        if case:
+          a = max(a, align(direction, case))
+      a
+    handle => align(i32)
+    push-buffer | pull-buffer => align(i32)
+```
+
+With those definitions in mind the way to lower types to memory is then
+relatively straightforward:
+
+```
+def lower_to_memory(direction, dst, ptr, value):
+  match value
+    f32(val) => write_4_bytes(dst, ptr, val),
+    f64(val) => write_8_bytes(dst, ptr, val),
+
+    s8(val) => write_1_byte(dst, ptr, val),
+    u8(val) => write_1_byte(dst, ptr, val),
+    s16(val) => write_2_bytes(dst, ptr, val),
+    u16(val) => write_2_bytes(dst, ptr, val),
+    s32(val) => write_4_bytes(dst, ptr, val),
+    u32(val) => write_4_bytes(dst, ptr, val),
+    s64(val) => write_8_bytes(dst, ptr, val),
+    u64(val) => write_8_bytes(dst, ptr, val),
+
+    char(val) => write_4_bytes(dst, ptr, val),
+
+    string(val) |
+    list<$ty>(val) => {
+      [val_ptr, val_len] = lower(direction, dst, val)
+      write_4_bytes(dst, ptr, val_ptr)
+      write_4_bytes(dst, ptr + 4, val_len)
+    }
+
+    bitflags<$fields>(val) => {
+      for i, bits in lower(direction, dst, val):
+        write_4_bytes(dst, ptr + i * 4, bits)
+    }
+
+    record<$fields>(val) => {
+      for $field in $fields:
+        ptr = align_to(ptr, align(direction, $field))
+        lower_to_memory(direction, dst, ptr, val[$field])
+        ptr += size(direction, $field)
+    }
+
+    variant<$cases>(val) => {
+      match val {
+        $case(payload) => {
+          # Write the discriminant, using the smallest size available
```

Comment:
The main thing I was trying to do with this was to match what WASI currently does which is roughly to shrink enum discriminants to the size  of the discriminant space. I figured that if WASI is doing something already it'd be nice to keep it in the canonical ABI to avoid clashes with "but WASI was faster?"

Reply:
Maybe we can forgo this optimization and add it back if anyone seems to especially care?  Otherwise, I think you'll also need to update `align` and `size`.